### PR TITLE
location_parent_id -> ancestor_location

### PR DESCRIPTION
**Pull request by sravfeyn**

### CODE REVIEW

**Review by emord — CHANGES_REQUESTED** [review]
Thanks for taking a look at this. Looks like there are some syntax issues here (which is why the tests failed)

Also I don't think that this will work. `ancestor_location` returns a json representation of the full location, so it needs to be either in a [`nested` expression](https://github.com/dimagi/commcare-hq/blob/9be7405b98afb5866a32e8aeb1b0198f51e04ce0/custom/icds_reports/ucr/data_sources/dashboard/delivery_forms.json#L31-L47) to extract the ID, or [have a `location_property`](https://github.com/dimagi/commcare-hq/blob/9be7405b98afb5866a32e8aeb1b0198f51e04ce0/custom/icds_reports/ucr/data_sources/dashboard/child_tasks.json#L40-L53)

The `location_property` is a shortcut I added a few days ago

**Review by emord — COMMENTED** [review]
There are still some syntax errors in the data sources here

File: custom/icds_reports/ucr/data_sources/awc_locations.json [hunk 1]
```diff
@@ -494,33 +494,41 @@
...
           "type": "named",
-          "name": "district_id"
+          "name": "awc_id"
         },
-        "type": "location_parent_id"
-      }
+        "location_type": "state",
+        "location_propery": "_id",
+        "type": "ancestor_location"
+      },
```
[the remark this answers is not in the document]
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/awc_locations.json]: Sorry, fixed it

**Comment by sravfeyn** [comment]
The tests are going okay locally, I think should be now good to go after build pass @emord

**Comment by sravfeyn** [comment]
@emord You have probably missed signing off on this. Is this still needed? I can resolve conflicts if needed

**Comment by emord** [comment]
@sravfeyn thanks, I totally lost track of this. I think that this would still be nice to have since it will reduce our time to process these UCRs

**Comment by sravfeyn** [comment]
@snopoke @emord I have fixed conflicts on this old PR that converts a whole bunch of location_parent_id expression to more efficient ancestor_location expression.

File: custom/icds_reports/ucr/data_sources/awc_mgt_forms.json [hunk 2]
```diff
@@ -87,11 +87,13 @@
...
         "expression": {
-          "location_id_expression": {
+          "location_id": {
             "type": "named",
             "name": "user_location_id"
           },
-          "type": "location_parent_id"
+          "location_type": "block",
+          "location_propery": "_id",
+          "type": "ancestor_location"
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/awc_mgt_forms.json]: I think this is only useful if you also change the expression for `state_id` otherwise the caching isn't re-used.

File: custom/icds_reports/ucr/data_sources/dashboard/delivery_forms.json [hunk 3]
```diff
@@ -35,15 +35,13 @@
         "expression": {
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/dashboard/delivery_forms.json]: the `nested` expression isn't needed here since you've added the `location_property` field. You can replace the entire expression value with the value of  `argument_expression`.

File: custom/icds_reports/ucr/data_sources/household_cases.json [hunk 4]
```diff
@@ -84,44 +85,38 @@
...
           },
-          "type": "location_parent_id"
+          "location_type": "block",
+          "type": "ancestor_location"
         },
         "transform": {},
         "is_nullable": true,
         "is_primary_key": false,
         "column_id": "block_id",
+        "create_index": false,
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/household_cases.json]: why was this added. The default is `false`

File: custom/icds_reports/ucr/data_sources/person_cases_v2.json [hunk 5]
```diff
@@ -66,16 +66,19 @@
         "is_nullable": true,
         "is_primary_key": false,
         "column_id": "supervisor_id",
+        "create_index": true,
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/person_cases_v2.json]: was adding this intentional?

File: custom/icds_reports/ucr/data_sources/tech_issue_cases.json [hunk 6]
```diff
@@ -112,36 +108,29 @@
...
-                "property_name": "ls_id"
-              }
+                "property_name": "aww_id"
             },
-            "type": "location_parent_id"
+            "expression_if_false": {
+                "type": "property_name",
+                "property_name": "ls_id"
+            }
           },
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/tech_issue_cases.json]: (nice to have) make this a named expression to avoid repetition

File: custom/icds_reports/ucr/data_sources/thr_forms_v2.json [hunk 7]
```diff
@@ -51,11 +51,13 @@
...
         "expression": {
-          "location_id_expression": {
+          "location_id": {
             "type": "named",
             "name": "user_location_id"
           },
-          "type": "location_parent_id"
+          "location_type": "supervisor",
+          "location_propery": "_id",
+          "type": "ancestor_location"
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/thr_forms_v2.json]: is this more efficient if there are no other expressions in the data source that also need it?

File: custom/icds_reports/ucr/data_sources/household_cases.json (hunk 4 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/household_cases.json]: I think it was a mistake, will remove it.

File: custom/icds_reports/ucr/data_sources/thr_forms_v2.json (hunk 7 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/thr_forms_v2.json]: Actually, looks like `ancestor_location` location makes two queries, one to get location and the other to get all ancestors. And `location_parent_id` makes only one query. So, it makes sense to use it if there is only 2 or more indicators needing `location_parent_id`

File: custom/icds_reports/ucr/data_sources/awc_mgt_forms.json (hunk 2 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/awc_mgt_forms.json]: Sure

File: custom/icds_reports/ucr/data_sources/tech_issue_cases.json (hunk 6 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/tech_issue_cases.json]: Done

File: custom/icds_reports/ucr/data_sources/thr_forms_v2.json (hunk 7 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/thr_forms_v2.json]: Updated the only two files with single uses to use `location_parent_id`

**Comment by sravfeyn** [comment]
@emord  @snopoke Addressed all comments, ran the files through jsonlint to make sure no syntax errors. Please can you guys do a review

**Comment by sravfeyn** [comment]
Build is passing now.

File: custom/icds_reports/ucr/data_sources/hardware_cases.json [hunk 8]
```diff
@@ -62,12 +62,17 @@
...
-            "datatype": null,
-            "type": "property_name",
-            "property_name": "owner_id"
+          "location_id": {
+            "expression": {
+              "datatype": null,
+              "type": "property_name",
+              "property_name": "owner_id"
+            },
+            "type": "root_doc"
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/hardware_cases.json]: this doesn't need to be a root doc expression, same below

Comment (snopoke) [reply custom/icds_reports/ucr/data_sources/hardware_cases.json]: I see a number of places where there are root doc expressions that aren't needed. Perhaps not worth changing for now.

File: custom/icds_reports/ucr/data_sources/tech_issue_cases.json [hunk 9]
```diff
@@ -353,31 +284,50 @@
...
+                "property_name": "aww_id"
+            },
+            "expression_if_false": {
+                "type": "property_name",
+                "property_name": "ls_id"
+            }
+          },
+          "location_type": "supervisor",
+          "location_propery": "_id",
+          "type": "ancestor_location"
```
Comment (snopoke) [inline custom/icds_reports/ucr/data_sources/tech_issue_cases.json]: I don't think this will work here. In the case of an LS the location won't be part of the ancestry and the expression doesn't consider the initial location. I think the best thing to do would be to move the `ancestor_location` expression to the `expression_if_true` clause.

File: custom/icds_reports/ucr/data_sources/person_cases_v2.json (hunk 5 above)
Comment (snopoke) [reply custom/icds_reports/ucr/data_sources/person_cases_v2.json]: @sravfeyn bump

**Review by snopoke — CHANGES_REQUESTED** [review]

File: custom/icds_reports/ucr/data_sources/tech_issue_cases.json (hunk 9 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/tech_issue_cases.json]: That's a good catch, fixed it.

File: custom/icds_reports/ucr/data_sources/hardware_cases.json (hunk 8 above)
Comment (sravfeyn) [reply custom/icds_reports/ucr/data_sources/hardware_cases.json]: Yeah, going to leave it for now

**Comment by sravfeyn** [comment]
Addressed latest comments as well @snopoke

**Review by snopoke — APPROVED** [review]

**Review by emord — APPROVED** [review]
